### PR TITLE
Compatibility with R-4.2

### LINKED ISSUE #374: Cannot built inline-r-0.10.5 with latest R version 4.2.0
Using R 4.1.3 i can `cabal build inline-r-0.10.5`.

Using latest R 4.2.0 (that also is on stable ArchLinux), I get on `cabal build inline-r-0.10.5`:

```
Configuring library for inline-r-0.10.5..
Preprocessing library for inline-r-0.10.5..
compiling dist/build/Foreign/R/Internal_hsc_make.c failed (exit code 1)
rsp file was: "dist/build/Foreign/R/hsc2hscall26684-0.rsp"
command was: /usr/bin/gcc -c dist/build/Foreign/R/Internal_hsc_make.c -o dist/build/Foreign/R/Internal_hsc_make.o -fuse-ld=gold -D__GLASGOW_HASKELL__=810 -Dlinux_BUILD_OS=1 -Dx86_64_BUILD_ARCH=1 -Dlinux_HOST_OS=1 -Dx86_64_HOST_ARCH=1 -Icbits -I/usr/include/R/ -Idist/build/cbits -I/usr/include/R/ -Idist/build/autogen -Idist/build/global-autogen -include dist/build/autogen/cabal_macros.h -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/process-1.6.13.2/include -I/home/mpla/.cabal/store/ghc-8.10.7/vector-0.12.3.1-5288858e421932da699ccb45be171a38682f0167fd3ea1835346124255781dd2/lib//include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/unix-2.7.2.2/include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/time-1.9.3/include -I/home/mpla/.cabal/store/ghc-8.10.7/primitive-0.7.3.0-ca993c0684c5b48074dd2590184c6f06667d73290dfde50301aed62da070af56/lib//include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/bytestring-0.10.12.0/include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/base-4.14.3.0/include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/integer-gmp-1.0.3.0/include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/include -I/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/include/
error: In file included from /usr/include/strings.h:23,
                 from /usr/include/string.h:462,
                 from /usr/include/R/R_ext/RS.h:34,
                 from /usr/include/R/R.h:80,
                 from Internal.hsc:46:
Internal.hsc: In function ‘main’:
Internal.hsc:142:15: error: unknown type name ‘SEXPREC’
Internal.hsc:142:5: note: in expansion of macro ‘hsc_poke’
Internal.hsc:146:15: error: unknown type name ‘SEXPREC’
Internal.hsc:146:5: note: in expansion of macro ‘hsc_poke’
Internal.hsc:150:15: error: unknown type name ‘SEXPREC’
Internal.hsc:150:5: note: in expansion of macro ‘hsc_poke’
Internal.hsc:195:15: error: unknown type name ‘VECTOR_SEXPREC’
Internal.hsc:195:5: note: in expansion of macro ‘hsc_peek’
In file included from dist/build/Foreign/R/Internal_hsc_make.c:1:
Internal.hsc:199:15: error: ‘SEXPREC_ALIGN’ undeclared (first use in this function)
/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/template-hsc.h:96:39: note: in definition of macro ‘hsc_size’
   96 |     hsc_printf("(%ld)", (long) sizeof(t));
      |                                       ^
Internal.hsc:199:15: note: each undeclared identifier is reported only once for each function it appears in
/home/mpla/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/template-hsc.h:96:39: note: in definition of macro ‘hsc_size’
   96 |     hsc_printf("(%ld)", (long) sizeof(t));
      |                                       ^
```

I am not sure if this is a bug of this Haskell library, a bug from the Archlinux packaging side, or a bug from the upstream R dev side.

### PR DESCRIPTION
**Pull request by mboes**
R-4.2 onwards no longer support exposing the memory layout of SEXP
forms by defining `USE_RINTERNALS`. So we can't rely on this knowledge
anymore and have to use accessor functions instead of direct field
access. The new code is probably a bit slower than what we did before.
But on the plus side we are likely more robust to future changes to
the R API.

Fixes #374

### CODE REVIEW

File: inline-r/src/Foreign/R.hsc [hunk 1]
```diff
@@ -192,14 +187,25 @@ C.include "<stdint.h>"
 car :: SEXP s a -> IO (SomeSEXP s)
 car (unsexp -> s) = somesexp <$> [C.exp| SEXP { CAR( $(SEXP s) ) } |]
 
+-- | Set the CAR value and return it.
+setCar :: SEXP s a -> SEXP s b -> IO (SEXP s b)
+setCar (unsexp -> s) (unsexp -> s') = sexp <$> [C.exp| SEXP { SETCAR( $(SEXP s), $(SEXP s') ) } |]
```
Comment (facundominguez) [inline inline-r/src/Foreign/R.hsc:192]: I'm guessing that you'd want to use `Language.C.Inline.Unsafe.exp` in these cases.

**Comment by idontgetoutmuch** [comment]
Thanks for doing this :-) in the meantime I have been writing out data as CSV and reading it back into R to draw my plots :-(

File: inline-r/src/Foreign/R.hsc (hunk 1 above)
Comment (mboes) [reply inline-r/src/Foreign/R.hsc:192]: Good point. Although that's true of a lot of functions in this module, none of which use `Language.C.Inline.Unsafe`. I suggest we address this in a followup.

File: inline-r/src/Language/R/HExp.hsc [hunk 2]
```diff
@@ -307,56 +307,60 @@ peekHExp s = do
...
+
+        su :: forall b. SEXP s b
+        su = R.unsafeCoerce s
 
     case R.typeOf s of
       R.Nil       -> coerce $ return Nil
       R.Symbol    -> coerce $
-        Symbol    <$> (coerceAny <$> R.sexp <$> #{peek SEXPREC, u.symsxp.pname} sptr)
-                  <*> (R.sexp <$> #{peek SEXPREC, u.symsxp.value} sptr)
-                  <*> (R.sexp <$> #{peek SEXPREC, u.symsxp.internal} sptr)
```
Comment (facundominguez) [inline inline-r/src/Language/R/HExp.hsc:337]: It looks like `HExp.hsc` doesn't use `hsc2hs` features and can also be promoted to a regular Haskell module now.

**Review by facundominguez — APPROVED** [review]
LGTM
Though I think something should be done to recover type safety when using `ccall`.

File: inline-r/src/Foreign/R/Internal.hsc [hunk 3]
```diff
@@ -148,6 +146,7 @@ setCdr s s' = #{poke SEXPREC, u.listsxp.cdrval} (unsexp s) (castPtr $ unsexp s')
 -- | Set TAG field of object, when object is viewed as a cons cell.
 setTag :: SEXP s a -> SEXP s b -> IO ()
 setTag s s' = #{poke SEXPREC, u.listsxp.tagval} (unsexp s) (castPtr $ unsexp s')
+foreign import ccall unsafe "TYPEOF" cTYPEOF :: SEXP0 -> IO CInt
```
[the remark this answers is not in the document]
Comment (facundominguez) [reply inline-r/src/Foreign/R/Internal.hsc:130]: ... or just use `inline-c` to bind these.

[the remark this answers is not in the document]
Comment (mboes) [reply inline-r/src/Foreign/R/Internal.hsc:130]: We could likely move all of them to `Foreign.R`, which already uses inline-c for everything. Except that we'd have to also move the `SEXPInfo` type. Say as part of merging `Foreign.R.Internal` wholesale into `Foreign.R`. The distinction isn't very useful anymore.

File: inline-r/src/Foreign/R/Internal.hsc [hunk 4]
```diff
@@ -122,28 +122,12 @@ cUIntFromSingEnum = cIntConv . fromEnum . fromSing
 cIntFromEnum :: Enum a => a -> CInt
 cIntFromEnum = cIntConv . fromEnum
 
---------------------------------------------------------------------------------
--- Generic accessor functions                                                 --
---------------------------------------------------------------------------------
-
```
[the remark this answers is not in the document]
Comment (mboes) [reply inline-r/src/Foreign/R/Internal.hsc:131]: Renamed.

File: inline-r/src/Language/R/HExp.hsc (hunk 2 above)
Comment (mboes) [reply inline-r/src/Language/R/HExp.hsc:337]: Renamed.

**Comment by andreabedini** [comment]
:tada:

**Pull request merged**